### PR TITLE
Add BERT Embedding Models

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/bert/loss.py [hunk 1]
```diff
@@ -96,6 +97,98 @@ def reduce(self, losses_reduced_per_micro_batch) -> torch.Tensor:
...
+        label_smoothing: float = 0.0,
+    ) -> None:
+        super().__init__()
+        self.validation_step = validation_step
+        self.val_drop_last = val_drop_last
+        self.num_hard_negatives = num_hard_negatives
+        self.scale = scale
+        self.cross_entropy_loss = nn.CrossEntropyLoss(label_smoothing=label_smoothing)
+
+    def forward(
```
Comment (JRD971000) [inline nemo/collections/llm/bert/loss.py:124]: I thin the current implementation only has in-batch negative for every rank, right? There is a an option to use all the global in-batch negative in nemo1, could you please add it:

https://github.com/NVIDIA/NeMo/blob/c35877c8e22d90919166752e069a9cf4f9ec0823/nemo/collections/nlp/models/information_retrieval/megatron_bert_embedding_model.py#L686

File: nemo/collections/llm/bert/model/embedding.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    _batch: dict
+    if isinstance(batch, tuple) and len(batch) == 3:
+        _batch = batch[0]
+    else:
+        _batch = batch
+
+    required_keys = set()
+    required_keys.add("attention_mask")
+    required_keys.add("token_type_ids")
+    # required_keys.add("metadata")
```
Comment (JRD971000) [inline nemo/collections/llm/bert/model/embedding.py:33]: please remove, sorry for nit-picking🫠

File: nemo/collections/llm/bert/data/fine_tuning.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            path,
+            tokenizer=self.tokenizer,
+            seq_length=self.seq_length,
+            memmap_workers=self.memmap_workers,
+            seed=self.seed,
+            **kwargs,
+        )
+
+    def _create_dataloader(self, dataset, mode, **kwargs) -> DataLoader:
+        # pylint: disable=C0115,C0116
```
Comment (JRD971000) [inline nemo/collections/llm/bert/data/fine_tuning.py:188]: all pylint comments are not required, right?

**Review by JRD971000 — CHANGES_REQUESTED** [review]
I think just one major comment for global in-batch negatives, the rest is removing commented codes and addressing the CodeQL comments. Thanks a lot Ao!

File: nemo/collections/llm/bert/data/fine_tuning.py (hunk 3 above)
Comment (suiyoubi) [reply nemo/collections/llm/bert/data/fine_tuning.py:188]: func starts with underscore indeed not required #pylint, removed those

**Review by JRD971000 — APPROVED** [review]
LGTM, Thanks Ao!

**Review by JRD971000 — APPROVED** [review]
